src/Connection.js

```diff
@@ -32,6 +32,14 @@ class Connection extends EventEmitter {
         } else if (this.state === Connection.State.CONNECTED) {
             return Promise.reject(new Error('Already connected!'))
         }
+        if (this.state === Connection.State.DISCONNECTING) {
+            return new Promise((resolve) => {
+                this.once('disconnected', async () => {
+                    await this.connect()
```

Comment:
if connect rejects, then I think this will deadlock. I think it should be: `resolve(this.connect())`. If connect rejects, then this will resolve to a rejection which is equivalent to calling `reject`

Reply:
also what should happen if there's an 'error' event before 'disconnected' happens? It should probably remove the once('disconnected' handler on error. Event emitters are fraught with peril.